wallet.js

```diff
@@ -176,73 +191,89 @@ function createWalletMiddleware(opts = {}) {
       message = firstParam
       address = secondParam
     }
+    address = await validateAndNormalizeKeyholder(address, req)
 
     const msgParams = Object.assign({}, extraParams, {
       from: address,
       data: message,
     })
-
-    await validateSender(address, req)
+    
     res.result = await processPersonalMessage(msgParams, req)
   }
 
   async function personalRecover(req, res) {
+
     const message = req.params[0]
     const signature = req.params[1]
-    // non-standard "extraParams" to be appended to our "msgParams" obj
     const extraParams = req.params[2] || {}
     const msgParams = Object.assign({}, extraParams, {
       sig: signature,
       data: message,
     })
+    const signerAddress = sigUtil.recoverPersonalSignature(msgParams)
 
-    const senderHex = sigUtil.recoverPersonalSignature(msgParams)
-    res.result = senderHex
+    res.result = signerAddress
   }
 
   async function encryptionPublicKey (req, res) {
-    if (!processEncryptionPublicKey) throw new Error('WalletMiddleware - opts.processEncryptionPublicKey not provided')
-    const address = req.params[0]
 
-    await validateSender(address, req)
+    if (!processEncryptionPublicKey) {
+      throw new Error('WalletMiddleware - opts.processEncryptionPublicKey not provided')
+    }
+
+    const address = await validateAndNormalizeKeyholder(req.params[0], req)
+
     res.result = await processEncryptionPublicKey(address, req)
   }
 	
   async function decryptMessage (req, res) {
-    if (!processDecryptMessage) throw new Error('WalletMiddleware - opts.processDecryptMessage not provided')
-    // process normally
-    const firstParam = req.params[0]
-    const secondParam = req.params[1]
-    // non-standard "extraParams" to be appended to our "msgParams" obj
-    const extraParams = req.params[2] || {}
 
+    if (!processDecryptMessage) {
+      throw new Error('WalletMiddleware - opts.processDecryptMessage not provided')
+    }
+
+    const ciphertext = req.params[0]
+    const address = await validateAndNormalizeKeyholder(req.params[1])
+    const extraParams = req.params[2] || {}
     const msgParams = Object.assign({}, extraParams, {
-      from: secondParam,
-      data: firstParam,
+      from: address,
+      data: ciphertext,
     })
 
-    await validateSender(secondParam, req)
     res.result = await processDecryptMessage(msgParams, req)
   }
 
   //
   // utility
   //
 
-  async function validateSender(address, req) {
-    // allow unspecified address (allow transaction signer to insert default)
```

Comment:
This seems to no longer be allowed. I'm not entirely sure yet which methods had allowed the "sender" address to be omitted, but requiring it all of a sudden seems like a breaking change.

Reply:
I was trying to figure out whether we actually inserted a "default" as the comment suggests but couldn't find any examples. I'm going to run some experiments and see if you can request any of these RPC methods without specifying a keyholder address.

Reply:
After testing `develop` and prod, I can confirm that we reject any request without a keyholder address for the wallet middleware RPC methods that require a keyholder address.

This includes `eth_sendTransaction`, where a default `from` address is injected by `web3` in our test dapp. We do not fallback to a default address in prod or `develop`.

Similarly, all signature methods fail in production if a keyholder address is not specified. Our UI handles this very poorly, so rejecting them in this middleware layer will circumvent that problem entirely.